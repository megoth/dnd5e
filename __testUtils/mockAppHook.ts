import mockApp from "./mockApp";

export default function mockAppHook(mock: jest.Mock, app = mockApp()) {
  mock.mockReturnValue(app);
}
