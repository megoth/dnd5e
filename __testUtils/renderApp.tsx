import React from "react";
import { render, RenderOptions, RenderResult } from "@testing-library/react";
import { LocalizationProvider, ReactLocalization } from "@fluent/react";
import { getFluentBundles } from "../src/models/language";
import { AppModel } from "../src/models/app";

export default function renderApp(
  app: AppModel,
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
): RenderResult {
  function rerender() {
    return render(
      <LocalizationProvider l10n={new ReactLocalization(getFluentBundles(app))}>
        {ui}
      </LocalizationProvider>,
      options
    );
  }
  const renderResult = rerender();
  return { ...renderResult, rerender };
}
