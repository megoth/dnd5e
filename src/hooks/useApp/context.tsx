import { createContext } from "react";
import { Locale } from "../../ldo/app.typings";

const AppContext = createContext<{
  currentLocale: string;
  availableLocales: Array<Locale>;
}>({
  currentLocale: "en-US",
  availableLocales: [],
});

export default AppContext;
