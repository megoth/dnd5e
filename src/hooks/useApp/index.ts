import { useContext, useEffect } from "react";
import { AppContext } from "../../contexts/app";
import { appIsLoadingLocalizations } from "../../models/app";

export default function useApp(bundles: string[] = []) {
  const { app, setBundles } = useContext(AppContext);
  useEffect(() => {
    if (!app || !appIsLoadingLocalizations(app, bundles)) {
      return;
    }
    setBundles(bundles);
  }, [app, bundles, setBundles]);
  return app;
}
