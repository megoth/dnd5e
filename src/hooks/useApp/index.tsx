import { useContext } from "react";
import AppContext from "./context";

export default function useApp() {
  return useContext(AppContext);
}
