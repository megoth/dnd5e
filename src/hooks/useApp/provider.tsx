import React, { ReactNode, useEffect, useMemo, useState } from "react";
import { LocalizationProvider, ReactLocalization } from "@fluent/react";
import { getLocale } from "../../utils/language";
import AppContext from "./context";
import useSWR from "swr";
import Loading from "../../components/loading";
import {
  AppShapeType,
  LocaleShapeType,
  ResourceBundleShapeType,
  TranslationShapeType,
  TranslationsIndexShapeType,
} from "../../ldo/app.shapeTypes";
import { SessionInfo, useLdo, useSolidAuth } from "@ldo/solid-react";
import { hash, resourceUrl } from "../../utils/url";
import { useSearchParams } from "react-router-dom";
import { FluentBundle, FluentResource } from "@fluent/bundle";
import { userIsAdmin } from "../../utils/session";
import { namedNode } from "@rdfjs/data-model";
import { useLocalStorage } from "@uidotdev/usehooks";
import { Literal } from "@rdfjs/types";

interface Props {
  children: ReactNode;
}

function getBundles(session: SessionInfo): string[] {
  return userIsAdmin(session) ? ["global", "admin"] : ["global"];
}

export default function AppProvider({ children }: Props) {
  const { session } = useSolidAuth();
  const { dataset, getResource, getSubject } = useLdo();
  const [searchParams] = useSearchParams();
  const [currentLocale, setCurrentLocale] = useLocalStorage("locale", "en-US");
  const [bundles, setBundles] = useState<string[]>(getBundles(session));

  useEffect(() => {
    setBundles(getBundles(session));
  }, [session]);

  // fetch app index
  const { data: app } = useSWR("app", async () => {
    const appUrl =
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      import.meta.env.VITE_APP_URL ||
      new URL("/data/index.ttl#dnd5e", location.origin).toString();
    const path = resourceUrl(appUrl);
    await getResource(path).readIfUnfetched();
    return getSubject(AppShapeType, appUrl);
  });

  // fetch resource bundles
  const { data: resourceBundles } = useSWR(
    () => `resourceBundles-${app["@id"]}}-${bundles.join("-")}`,
    async () =>
      (
        await Promise.all(
          app.resourceBundle.map(async (bundle) => {
            await getResource(bundle["@id"]).readIfUnfetched();
            return getSubject(ResourceBundleShapeType, bundle["@id"]);
          }),
        )
      ).filter((bundle) => bundles.indexOf(bundle.label) !== -1),
  );

  // fetch supported languages
  const { data: supportedLocales } = useSWR(
    () => `supportedLocales-${app["@id"]}}`,
    async () => {
      return Promise.all(
        app.supportLanguage.map(async (language) => {
          await getResource(language["@id"]).readIfUnfetched();
          return getSubject(LocaleShapeType, language["@id"]);
        }),
      );
    },
  );

  useEffect(() => {
    if (!supportedLocales) return;
    // TODO: FIX so that you can load translations directly
    const requestedLocale = searchParams.get("locale");
    const locale = getLocale(
      (Array.isArray(requestedLocale) ? requestedLocale[0] : requestedLocale) ||
        currentLocale,
      supportedLocales?.map((locale) => locale.language),
    );
    setCurrentLocale(locale);
  }, [searchParams.get("locale"), supportedLocales, currentLocale]);

  // fetch FAQs
  const { isLoading: faqIsLoading } = useSWR(
    () => `faqs-${resourceBundles.map((b) => b["@id"]).join("-")}}`,
    async () =>
      Promise.all(
        resourceBundles
          .flatMap((bundle) => bundle.faqIndex)
          .map(async (index) => getResource(index["@id"]).readIfUnfetched()),
      ),
  );

  // fetch translation indices
  const { data: translations } = useSWR(
    () =>
      `translationsIndices-${resourceBundles.map((b) => b["@id"]).join("-")}-${currentLocale}`,
    async () => {
      return (
        await Promise.all(
          resourceBundles
            .flatMap((bundle) => bundle.translationsIndex)
            .map(async (index) => {
              await getResource(index["@id"]).readIfUnfetched();
              return getSubject(TranslationsIndexShapeType, index["@id"]);
            }),
        )
      ).filter((index) => !index.language || index.language === currentLocale);
    },
  );

  // get translations
  const { isLoading: translationsLoading } = useSWR(
    () => `translations-${translations.map((t) => t["@id"]).join("-")}}`,
    async () =>
      Promise.all(
        translations
          .map((index) => index.translationsResource)
          .map(async (resourceUrl) => {
            return getResource(resourceUrl["@id"]).readIfUnfetched();
          }),
      ),
  );
  const localization = useMemo(() => {
    const localization = new FluentBundle(currentLocale);
    if (!resourceBundles || translationsLoading) return null;
    resourceBundles.forEach((bundle) =>
      bundle.translationsIndex
        .map((index) => index.translationsResource["@id"])
        .flatMap((resourceUrl) => {
          return dataset
            .match(null, null, null, namedNode(resourceUrl))
            .toArray()
            .filter(
              (quad) =>
                (quad.object as Literal).language ===
                currentLocale.toLowerCase(),
            )
            .map((quad) =>
              getSubject(TranslationShapeType, quad.subject.value),
            );
        })
        .forEach((translation) =>
          localization.addResource(
            new FluentResource(
              `${hash(translation["@id"])} = ${translation.definition}`,
            ),
          ),
        ),
    );
    return new ReactLocalization([localization]);
  }, [currentLocale, resourceBundles, dataset, translationsLoading]);

  if (translationsLoading || faqIsLoading || !translations || !localization) {
    return <Loading />;
  }

  return (
    <AppContext.Provider
      value={{
        app,
        currentLocale,
        availableLocales: supportedLocales?.filter(
          (locale) => locale.language !== currentLocale,
        ),
      }}
    >
      <LocalizationProvider l10n={localization}>
        {children}
      </LocalizationProvider>
    </AppContext.Provider>
  );
}
