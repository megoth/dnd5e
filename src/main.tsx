import React, {
  lazy,
  type LazyExoticComponent,
  ReactElement,
  ReactNode,
  Suspense,
} from "react";
import ReactDOM from "react-dom/client";
import { createBrowserRouter, Outlet, RouterProvider } from "react-router-dom";
import { BrowserSolidLdoProvider } from "@ldo/solid-react";
import HomePage from "./components/homePage";
import AppProvider from "./hooks/useApp/provider";
import "./styles/globals.css";
import FAQPage from "./components/faqPage";
import CharactersPage from "./components/charactersPage";
import ClassesPage from "./components/classesPage";
import EquipmentIndexPage from "./components/equipmentIndexPage";
import MonstersPage from "./components/monstersPage";
import RacesPage from "./components/racesPage";
import SpellsPage from "./components/spellsPage";
import LoginPage from "./components/loginPage";
import SignupPage from "./components/signupPage";
import RulesPage from "./components/rulesPage";
import LayoutProvider from "./hooks/useLayout/provider";
import ClassPage from "./components/classPage";
import SpellPage from "./components/spellPage";
import SearchPage from "./components/searchPage";
import RacePage from "./components/racePage";
import WeaponsPage from "./components/weaponsPage";
import ArmorIndexPage from "./components/armorIndexPage";
import MonsterPage from "./components/monsterPage";
import WeaponPage from "./components/weaponPage";
import ArmorSinglePage from "./components/armorSinglePage";
import EquipmentSinglePage from "./components/equipmentSinglePage";
import SubclassesPage from "./components/subclassesPage";
import SubracesPage from "./components/subracesPage";
import MagicItemsPage from "./components/magicItemsPage";
import MagicItemPage from "./components/MagicItemPage";
import YouPage from "./components/youPage";
import StoragesPage from "./components/storagesPage";
import StorageCreatePage from "./components/storageCreatePage";
import StoragePage from "./components/storagePage";
import CharacterCreatePage from "./components/characterCreatePage";
import SettingsPage from "./components/settingsPage";
import Loading from "./components/loading";

function lazyLoadPage(modulePath: string) {
  const Component = lazy(() => import(modulePath));
  return (
    <Suspense fallback={<Loading />}>
      <Component />
    </Suspense>
  );
}

const router = createBrowserRouter([
  {
    path: "/",
    element: (
      <BrowserSolidLdoProvider>
        <AppProvider>
          <LayoutProvider>
            <Outlet />
          </LayoutProvider>
        </AppProvider>
      </BrowserSolidLdoProvider>
    ),
    children: [
      {
        index: true,
        element: <HomePage />,
      },
      {
        path: "/about",
        element: ((Component) => {
          return (
            <Suspense fallback={<Loading />}>
              <Component />
            </Suspense>
          );
        })(lazy(() => import("./components/aboutPage"))),
      },
      {
        path: "/armor",
        element: <ArmorIndexPage />,
      },
      {
        path: "/armor/:url",
        element: <ArmorSinglePage />,
      },
      {
        path: "/characters",
        element: <CharactersPage />,
      },
      {
        path: "/characters/create",
        element: <CharacterCreatePage />,
      },
      {
        path: "/faq",
        element: lazyLoadPage("./components/faqPage"),
      },
      {
        path: "/login",
        element: <LoginPage />,
      },
      {
        path: "/rules",
        element: <RulesPage />,
      },
      {
        path: "/classes",
        element: <ClassesPage />,
      },
      {
        path: "/classes/:url",
        element: <ClassPage />,
      },
      {
        path: "/equipment",
        element: <EquipmentIndexPage />,
      },
      {
        path: "/equipment/:url",
        element: <EquipmentSinglePage />,
      },
      {
        path: "/magic-items",
        element: <MagicItemsPage />,
      },
      {
        path: "/magic-items/:url",
        element: <MagicItemPage />,
      },
      {
        path: "/monsters",
        element: <MonstersPage />,
      },
      {
        path: "/monsters/:url",
        element: <MonsterPage />,
      },
      {
        path: "/races",
        element: <RacesPage />,
      },
      {
        path: "/races/:url",
        element: <RacePage />,
      },
      {
        path: "/search",
        element: <SearchPage />,
      },
      {
        path: "/settings",
        element: <SettingsPage />,
      },
      {
        path: "/signup",
        element: <SignupPage />,
      },
      {
        path: "/spells",
        element: <SpellsPage />,
      },
      {
        path: "/spells/:url",
        element: <SpellPage />,
      },
      {
        path: "/storages",
        element: <StoragesPage />,
      },
      {
        path: "/storages/create",
        element: <StorageCreatePage />,
      },
      {
        path: "/storages/:url",
        element: <StoragePage />,
      },
      {
        path: "/subclasses",
        element: <SubclassesPage />,
      },
      {
        path: "/subraces",
        element: <SubracesPage />,
      },
      {
        path: "/weapons",
        element: <WeaponsPage />,
      },
      {
        path: "/weapons/:url",
        element: <WeaponPage />,
      },
      {
        path: "/you",
        element: <YouPage />,
      },
    ],
  },
]);

ReactDOM.createRoot(document.getElementById("root")!).render(
  <React.StrictMode>
    <RouterProvider router={router} />
  </React.StrictMode>,
);
