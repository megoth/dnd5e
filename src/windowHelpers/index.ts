/* istanbul ignore file */
// eslint-disable-next-line import/prefer-default-export
export function getRedirectURL(path) {
  if (typeof window !== "undefined") {
    const currentOrigin = window.location.origin;
    return `${currentOrigin}/${path}`;
  }

  return "";
}

export function getNavigatorLanguages() {
  if (typeof navigator === "undefined") {
    return ["en-US"];
  }
  return [...navigator.languages];
}
