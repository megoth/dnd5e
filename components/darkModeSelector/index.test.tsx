import React from "react";
import userEvent from "@testing-library/user-event";
import useApp from "../../src/hooks/useApp";
import mockAppHook from "../../__testUtils/mockAppHook";
import DarkModeSelector, { TESTID_DARK_MODE_SELECTOR_BUTTON } from "./index";
import * as windowFns from "../../src/windowHelpers";
import renderApp from "../../__testUtils/renderApp";

jest.mock("../../src/hooks/useApp");
const mockedAppHook = useApp as jest.Mock;

describe("DarkModeSelector", () => {
  let app;
  let mockedPrefersDarkModeSchema;

  beforeEach(() => {
    app = mockAppHook(mockedAppHook);
  });
  beforeEach(() => {
    mockedPrefersDarkModeSchema = jest
      .spyOn(windowFns, "prefersDarkModeScheme")
      .mockReturnValue(false);
  });

  it("renders", () => {
    const { asFragment } = renderApp(app, <DarkModeSelector />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders for dark mode", () => {
    mockedPrefersDarkModeSchema.mockReturnValue(true);
    const { asFragment } = renderApp(app, <DarkModeSelector />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("allows setting dark mode explicitly", () => {
    const { getByTestId } = renderApp(app, <DarkModeSelector />);
    const button = getByTestId(TESTID_DARK_MODE_SELECTOR_BUTTON);

    userEvent.click(button);
    expect(
      document.querySelector("html").classList.contains("dark")
    ).toBeTruthy();
    expect(localStorage.getItem("darkMode")).toBe("true");

    userEvent.click(button);
    expect(
      document.querySelector("html").classList.contains("dark")
    ).toBeFalsy();
    expect(localStorage.getItem("darkMode")).toBe("false");
  });
});
