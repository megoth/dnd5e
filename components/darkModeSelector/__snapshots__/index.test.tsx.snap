// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DarkModeSelector renders 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="dark-mode-selector-button"
    type="button"
  >
    <span>
      ðŸŒ™Â [Translation for https://example.com/global/translations.ttl#darkModeTurnOn does not exist]
    </span>
  </button>
</DocumentFragment>
`;

exports[`DarkModeSelector renders for dark mode 1`] = `
<DocumentFragment>
  <button
    class="button"
    data-testid="dark-mode-selector-button"
    type="button"
  >
    <span>
      ðŸ’¡Â [Translation for https://example.com/global/translations.ttl#darkModeTurnOff does not exist]
    </span>
  </button>
</DocumentFragment>
`;
